desc/compute/_bootstrap.py

```diff
@@ -0,0 +1,504 @@
+    be true in the case of circular cross-section surfaces in
+    axisymmetry with :math:`B \propto 1/R` and :math:`R = (1 +
+    \epsilon \cos\theta) R_0`.
+
+    This function is not formulated as a DESC "compute" function in
+    anticipation that it will be used with multiple methods of
+    computing the "nearby perfectly quasisymmetric field" in fields
+    that are only approximately quasisymmetric.
+
+    This function returns a Dictionary containing the following data,
+    all 1D arrays of shape ``(grid.num_rho,)``:
+    - ``"Bmin"``: The minimum of :math:`|B|` on each surface.
+    - ``"Bmax"``: The maximum of :math:`|B|` on each surface.
+    - ``"epsilon"``: The effective inverse aspect ratio on each surface.
+    - ``"<1/B>"``: :math:`\left<B^2\right>` on each surface,
+      where :math:`\left< \ldots \right>` denotes a flux surface average.
+    - ``"<1/B>"``: :math:`\left<1/B\right>` on each surface,
+      where :math:`\left< \ldots \right>` denotes a flux surface average.
+    - ``"f_t"``: The effective trapped fraction on each surface.
+
+    Parameters
+    ----------
+    grid : A ``Grid`` object.
+    modB : Magnetic field strength :math:`|B|` on the grid points.
+    sqrt_g : The Jacobian :math:`1/(\nabla\rho\times\nabla\theta\cdot\nabla\zeta)`
+        on the grid points.
+    n_gauss : int
+        Number of Gauss-Legendre integration points for the :math:`\lambda` integral.
+
+    Returns
+    -------
+    f_t_data : dict
+        Dictionary containing the computed data listed above.
+    """
+    denominator = surface_integrals(grid, sqrt_g)
+    fsa_B2 = compress(
+        grid, surface_averages(grid, modB * modB, sqrt_g, denominator=denominator)
+    )
+    fsa_1overB = compress(
+        grid, surface_averages(grid, 1 / modB, sqrt_g, denominator=denominator)
+    )
+
+    Bmax = surface_max(grid, modB)
+    Bmin = surface_min(grid, modB)
+    w = Bmax / Bmin
+    epsilon = (w - 1) / (w + 1)
+
+    # Get nodes and weights for Gauss-Legendre integration:
+    base_nodes, base_weights = roots_legendre(n_gauss)
+    # Rescale for integration on [0, 1], not [-1, 1]:
+    lambd = (base_nodes + 1) * 0.5
+    lambda_weights = base_weights * 0.5
+
+    modB_over_Bmax = modB / expand(grid, Bmax)
+
+    # Sum over the lambda grid points, using fori_loop for efficiency.
+    lambd = jnp.asarray(lambd)
+    lambda_weights = jnp.asarray(lambda_weights)
+
+    def body_fun(jlambda, lambda_integral):
+        flux_surf_avg_term = surface_averages(
+            grid,
+            jnp.sqrt(1 - lambd[jlambda] * modB_over_Bmax),
+            sqrt_g,
+            denominator=denominator,
+        )
+        return lambda_integral + lambda_weights[jlambda] * lambd[jlambda] / (
+            Bmax * Bmax * compress(grid, flux_surf_avg_term)
+        )
+
+    lambda_integral = fori_loop(0, n_gauss, body_fun, jnp.zeros(grid.num_rho))
+
+    f_t = 1 - 0.75 * fsa_B2 * lambda_integral
+    f_t_data = {
```

Comment:
Yeah, I was debating whether to add the trapped fraction, Bmax, etc to the data index. Here's the reason it's not done at the moment. There are 2 ways I can think of to compute these quantities for the Redl bootstrap current formula when a config is not perfectly quasisymmetric. In method 1, the one implemented in this PR, Bmax, Bmin, and f_t are computed from the full not-quite-QS field. In method 2, which is not in this PR but which I'm hoping to try soon, the B field is transformed to Boozer coordinates and non-QS modes are filtered out before evaluating Bmax, Bmin, and f_t. I expect method 2 will be more accurate but slower / more memory needed, but haven't had a chance to check. Methods 1 and 2 would re-use all the code in `trapped_fraction()`, but use different input values for `modB` and `sqrt_g`, and give different answers for `f_t`, `Bmax`, etc. 

Some possible ways to proceed:

- Leave things as they are for now. I can try out Method 2 in a branch, and see which of the two methods works better. Then we re-assess whether to refactor `trapped_fraction()` into several regular `compute` functions.
- I could refactor `trapped_fraction()` into several regular `compute` functions now. When I try out Method 2 (in a branch), there might be a little bit of repeated code, and there would be 2 variants of these quantities in the data index, e.g. `f_t` and `f_t Boozer filtered`, `Bmax` and `Bmax Boozer filtered`, etc.
- Other ideas welcome too.

Reply:
I would vote for the 2nd option. If the boozer filtered method does seem significantly better (and not too slow) we could refactor things to always use that, otherwise I think having 2 versions could be useful. My hunch is that the boozer filtering will be quite a bit slower so in many cases we may still want to use the current implementation.

Reply:
Sounds good. This function has now been refactored and split up into standard compute functions, so `trapped fraction` and intermediate quantities are all added to the data index.